normalize_and_clean_data.py

```diff
@@ -0,0 +1,204 @@
+        for correlated_column in correlated_columns:
+            df[column] = df[column].fillna(df[correlated_column])
+        # Якщо ще залишились пропуски, заповнюємо середнім значенням у межах міста
+        df[column] = df.groupby('city')[column].transform(lambda x: x.fillna(x.mean()))
+    logging.info('Пропущені значення відстаней заповнено')
+    return df
+
+def fill_missing_has_elevator(df: pd.DataFrame) -> pd.DataFrame:
+    logging.info('Заповнення пропусків у стовпці hasElevator')
+    df['hasElevator'] = df.apply(
+        lambda row: 'yes' if pd.isna(row['hasElevator']) and row['floorCount'] > 5 else ('no' if pd.isna(row['hasElevator']) else row['hasElevator']),
+        axis=1
+    )
+    logging.info('Пропущені значення hasElevator заповнено залежно від floorCount')
+    return df
+
+def replace_yes_no(df: pd.DataFrame) -> pd.DataFrame:
+    '''Заміна значень "yes" на 1 та "no" на 0 у всіх стовпцях'''
+    logging.info('Заміна значень yes/no на 1/0 у всіх стовпцях')
+    df = df.replace({'yes': 1, 'no': 0})
+    return df
+
+def replace_ownership_value(df: pd.DataFrame) -> pd.DataFrame:
+    logging.info('Заміна значення udział на part ownership у стовпці ownership')
+    df['ownership'] = df['ownership'].replace('udział', 'part ownership')
+    return df
+
+def save_clean_data(df: pd.DataFrame, save_directory: str) -> None:
+    logging.info('Збереження очищених даних')
+    # Створюємо назву файлу з датою та часом
+    current_datetime: str = datetime.now().strftime('%Y%m%d_%H%M%S')
+    file_name: str = f"cleaned_data_{current_datetime}.csv"
+    save_path: str = os.path.join(save_directory, file_name)
+    # Зберігаємо очищені дані у вказаній папці
+    df.to_csv(save_path, index=False)
+    logging.info('Очищені дані збережено за адресою: %s', save_path)
+
+def check_missing_values(df: pd.DataFrame) -> pd.DataFrame:
+    logging.info('Перевірка залишкових пропусків у даних')
+    # Перевірка, чи залишились пропуски після заповнення
+    missing_summary: pd.DataFrame = df.isnull().sum().to_frame(name='Missing Values')
+    missing_summary['Percentage'] = (missing_summary['Missing Values'] / len(df)) * 100
+    logging.info('Остаточний звіт про пропуски: %s', missing_summary)
+    print("Остаточний звіт про пропуски:")
+    print(missing_summary)
+    return df
+
+def main() -> None:
+    logging.info('-----Start-----')
+    directory_path: str = r'.\raw_data'
+    clean_folder: str = r'.\clean_data'
+
+    # Завантаження даних
+    df: pd.DataFrame = load_all_data(directory_path)
+    df = remove_non_unique_ids(df)
+    df = fill_condition(df)
+    df = fill_missing_building_material(df)
+    df = fill_missing_type(df)
+    df = fill_missing_build_year(df)
+    df = fill_missing_floor_count(df)
+    df = fill_missing_floor(df)
+    df = fill_missing_distances(df)
+    df = fill_missing_has_elevator(df)
+    df = replace_yes_no(df)
+    df = replace_ownership_value(df)
```

Comment:
Check if this methods inplace. https://www.geeksforgeeks.org/inplace-vs-standard-operators-python/.

I am not shure but it seams like you overrite df which changed values to same values. Check it!

Reply:
+